### PR TITLE
Be explict about some i18n oddness around undefined keys, and TextField's value

### PR DESCRIPTION
**Pull request by BPScott**
### WHY are these changes introduced?

Following up https://github.com/Shopify/polaris-react/pull/2579.

### WHAT is this pull request doing?

Make passing `{foo: undefined}` into translations replacements a runtime error, in addition to a TS type error for those fun cases where people power through mistyping things using `any` (looking at you enzyme's trigger function that isn't type-safe).

Make TextField's `normalizedValue` be an empty string if you pass anything other than a string into it. Once again this is limited to be a string only by typescript but people using those not-type-safe functions can pass in a number and not be warned. Having normalizedValue be a number mucks up a bunch of things so catch that at runtime too.

Make TextField a little lazier by only working out character count stuff if we're actually about to display it. 

### How to 🎩

tests pass

### CODE REVIEW

File: src/components/TextField/TextField.tsx [hunk 1]
```diff
@@ -181,7 +181,10 @@ export function TextField({
     focused ? input.focus() : input.blur();
   }, [focused]);
 
-  const normalizedValue = value != null ? value : '';
+  // Use a typeof check here as Typescript mostly protects us from non-stringy
+  // values but overzealous usage of `any` in consuming apps means people have
+  // been known to pass a number in, so make it clear that doesn't work.
+  const normalizedValue = typeof value === 'string' ? value : '';
```
Comment (dleroux) [inline src/components/TextField/TextField.tsx]: I like this approach.

File: src/utilities/i18n/I18n.ts [hunk 2]
```diff
@@ -33,22 +33,15 @@ export class I18n {
...
           );
         }
 
-        // Bringing back an old bit of oddness to expedite a release.
-        // {foo: undefined} is allowed in calling apps but won't trigger type
-        // warnings and won't get caught in the above check so let it through
-        // for now as JS can handle it ok. We should work out how to be
-        // stricter, or deliberatly allow undefined as a value
-        return replacements[replacement] as string;
+        return replacements[replacement].toString();
```
Comment (dleroux) [inline src/utilities/i18n/I18n.ts]: 👍

File: src/utilities/i18n/tests/I18n.test.ts [hunk 3]
```diff
@@ -35,7 +35,16 @@ describe('I18n.translate', () => {
     expect(() => {
       i18n.translate('key', {notString: 'bar'});
     }).toThrow(
-      `No replacement found for key 'string'. The following replacements were passed: 'notString'`,
+      `Error in translation for key 'key'. No replacement found for key 'string'. The following replacements were passed: '{"notString":"bar"}'`,
+    );
+
```
Comment (dleroux) [inline src/utilities/i18n/tests/I18n.test.ts]: Not 100% necessary but it might be best to split these 2 expect statements into 2 tests?

**Comment by BPScott** [comment]
A bit or rebasing to split these changes into isolated commits

**Pull request merged**